docs/source/hallway.rst

```diff
@@ -0,0 +1,363 @@
+------------------------------------
+
+Before we jump into the graph formulation of the hallway problem, let’s
+talk first about *actions*, because this is one of the key differences
+between ``graphenv`` and traditional RL gym environments.
+
+Typically, discrete actions spaces in gym environments have fixed
+dimension and fixed definitions. In the hallway problem, for instance,
+there are always two actions: “move left” (action=0) and “move right”
+(action=1). Cases where the action is infeasible, like trying to move
+left from the start position, are typically handled by implmenting a
+*null* transition where the action doesn’t change the current state.
+
+In general graph search problems, however, such fixed action spaces are
+not practical. In the game of chess, for example, the total number of
+possible board states and, subsequently, moves, is astronomical, while
+the set of *feasible* moves changes continually throughout the game.
+
+The key observation that makes graph search tractable as a gym
+environment is: even for large and complex graph search problems, the
+number of states that are accessible from the current state is usually
+relatively small. If, instead of fixing the action to a pre-defined set
+of choices, we think of each action as representing an accessible next
+state, this endeavor becomes tractable.
+
+And so, we abandon the idea of “fixed” action spaces in favor of
+“parametric” action spaces. Here, **parametric** means that actions
+coincide with next states represented by feature vectors, rather than
+having a single, index-based interpretation. In other words, rather than
+actions as “move :math:`[ left, right ]`”, parametric actions can be
+thought of as “go to state :math:`[i, j]`” where the states :math:`i,j`
+each have a vector representation. Also unliked fixed spaces, the order
+doesn’t matter: we could equivalently say “go to state :math:`[j, i]`”.
+
+A key ingredient in making this machinery work is to have policy models
+that can work on parameteric action spaces
+
+The figure below illustrates how to think of the simple hallway example
+as a graph problem.
+
+.. figure:: ./img/hallway-graph.png
+   :alt: hallway-graph
+
+   hallway-graph
+
+Before jumping into the specifics of how all of this works in
+``graphenv``, let’s define some terms.
+
+A **vertex** reprents a single state in the graph which, for this
+problem, can be described by an index :math:`i\in \{ 0, 1, 2 \}`.
+(Sometimes we’ll use the terms vertex, state, and node interchangeably).
+In the figure, each vertex is shown alongside the corresponding state of
+the hallway problem.
+
+The **root** is the starting vertex of the graph search, here,
+:math:`i=0`.
+
+At each state in the search, a certain number of child states (or
+**children**) are accessible. In the figure above, we illustrate this
+using the color codes:
+
+-  black = current vertex
+-  white = child vertex
+-  gray = inaccessible, or **masked**, vertex
```

Comment:
The masked vertex here is a bit odd -- it's not like we maintain a single global list of possible vertices and mask those that aren't reachable. For the molecule stuff (and probably other problems), we probably couldn't enumerate the whole graph *a priori* anyways.

Do you use the ![](https://latex.codecogs.com/svg.image?r_{01}) notation later? Another concept we might want to discuss (perhaps not here) is that we're really handling some sort of acyclic Markov chain: the only inputs to a given node's value is it's observation, unlike the (state, action) embeddings that would translate here as (parent, child) observations.

Although we do have some support for intermediate rewards, right? So we're not completely history-agnostic when it comes to evaluating paths through the search tree.

Reply:
agreed, will stick with "inaccessible"

for the r_01 notation, yeah i was on the fence about this because from an implementation standpoint, we just attach the reward to a vertex.  but from the standpoint of the search, it is a collected during the transition from i->j -- will use this instead, esp since is more standard RL formulation of (state + action) -> reward